### PR TITLE
Follow and watch settings.json single-level symlinks

### LINKED ISSUE #8547: Investigate why configuration via symlinks no longer update
- VSCode Version: Code - Insiders 1.3.0-insider (798b41ab1cf3636a27042020a54ddfb4cfb8d23c, 2016-06-29T05:14:50.570Z)
- OS Version: Linux x64 4.4.0-24-generic, Ubuntu 16.04

Steps to Reproduce:

1. Open settings
2. Change a setting that should have an immediate impact, notice there is no change

I have a feeling this is because I have my settings.json symlinked and it's not being watched correctly.

```
❯ ls -la ~/.config/Code\ -\ Insiders/User
total 16
drwx------ 4 daniel daniel 4096 Jun 23 22:46 .
drwx------ 8 daniel daniel 4096 Jun 27 03:38 ..
drwx------ 2 daniel daniel 4096 Mar  2 10:58 snippets
drwxrwxr-x 5 daniel daniel 4096 Jun 27 02:24 workspaceStorage
lrwxrwxrwx 1 daniel daniel   47 May 21 16:20 keybindings.json -> /home/daniel/.config/Code/User/keybindings.json
lrwxrwxrwx 1 daniel daniel   44 May 21 16:20 settings.json -> /home/daniel/.config/Code/User/settings.json
```

### PR DESCRIPTION
**Pull request by Tyriar**
Fixes #8547

---

Not sure how it ever worked looking at the commit history, this works though. Couldn't find an option to give to the watcher.

### CODE REVIEW

**Comment by joaomoreno** [comment]
I'd leave this for @bpasero. He won't be back for a couple of weeks though.

**Comment by bpasero** [comment]
@Tyriar as far as I can tell from your change you change the watching behavior to only watch on settings.json and not the parent folder. Did you try that other settings like keybindings are still being watched for changes?

**Comment by Tyriar** [comment]
@bpasero do you know where the keybindings are being watched, it's mentioned in this file but a watcher doesn't seem to exist? When I did this originally I assumed keybindings needed a full restart to get working, I don't think they ever worked with symlinks.

**Comment by bpasero** [comment]
@Tyriar we watch the parent folder of the settings file so every change within causes the event to trigger and keybindings is next to storage.json as far as I know. You do NOT need a restart for changing keybindings, wtf :)

**Comment by Tyriar** [comment]
@bpasero yep, I realized that a few weeks ago lol. Is `keybindings.json` the only other file we want this change event firing for?

**Comment by bpasero** [comment]
I believe so, but since this file is now being used in other processes I am not sure what else ends up in there. I think it would be safer to install the watcher on the folder, read the files inside and for each file that is a link to another file, install the watcher on that resolved file.

**Comment by Tyriar** [comment]
I like it :+1:

**Comment by Tyriar** [comment]
@bpasero ready for review, I tested using symlink and regular settings/keybindings files.

File: src/vs/base/node/userSettings.ts [hunk 1]
```diff
@@ -90,8 +90,35 @@ export class UserSettings {
 	}
 
 	private registerWatchers(): void {
-		this.watcher = fs.watch(path.dirname(this.appSettingsPath));
-		this.watcher.on('change', (eventType: string, fileName: string) => this.onSettingsFileChange(eventType, fileName));
+		let self = this;
+		function attachSettingsChangeWatcher(watchPath: string): void {
+			self.watcher = fs.watch(watchPath);
+			self.watcher.on('change', (eventType: string, fileName: string) => self.onSettingsFileChange(eventType, fileName));
```
Comment (bpasero) [inline src/vs/base/node/userSettings.ts]: I think onSettingsFileChange ignores eventType and fileName so maybe just remove it from there and invocations.

File: src/vs/base/node/userSettings.ts [hunk 2]
```diff
@@ -90,8 +90,35 @@ export class UserSettings {
...
+			this.appSettingsPath,
+			this.appKeybindingsPath
+		];
+		followSymlinkPaths.forEach((path) => {
+			fs.lstat(path, function(err, stats) {
+				if (err) {
+					return;
+				}
+				if (stats.isSymbolicLink()) {
+					fs.readlink(path, function(err, realPath) {
```
Comment (bpasero) [inline src/vs/base/node/userSettings.ts]: What if the symlink is actually a folder? Can you check that the stat is a file?

**Comment by bpasero** [comment]
@Tyriar LGTM but added 2 comments.

**Pull request merged**